File: backend/app/schemas/calificaciones.py
"""Schemas Pydantic para calificaciones y umbral de materia."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


# ── Calificacion ────────────────────────────────────────────────


class CalificacionCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    entrada_padron_id: UUID
    materia_id: UUID
    actividad: str
    nota_numerica: float | None = None
    nota_textual: str | None = None


class CalificacionResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True)

    id: UUID
    tenant_id: UUID
    entrada_padron_id: UUID
    materia_id: UUID
    actividad: str
    nota_numerica: float | None = None
    nota_textual: str | None = None
    aprobado: bool
    origen: str
    importado_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None = None


class CalificacionListResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    items: list[CalificacionResponse]
    total: int


# ── UmbralMateria ───────────────────────────────────────────────


class UmbralMateriaCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    umbral_pct: int = 60
    valores_aprobatorios: list[str] = ["Satisfactorio", "Supera lo esperado"]


class UmbralMateriaUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    umbral_pct: int | None = None
    valores_aprobatorios: list[str] | None = None


class UmbralMateriaResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True)

    id: UUID
    tenant_id: UUID
    asignacion_id: UUID
    materia_id: UUID
    umbral_pct: int
    valores_aprobatorios: list[str] | None = None
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None = None


# ── Import Preview ──────────────────────────────────────────────


class DetectedActivity(BaseModel):
    model_config = ConfigDict(extra="forbid")

    nombre: str
    tipo: str  # "numeric" | "textual"


class StudentMatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    entrada_padron_id: UUID
    nombre: str
    apellidos: str
    email: str
    datos: dict[str, str | float | None]


class UnmatchedRow(BaseModel):
    model_config = ConfigDict(extra="forbid")

    fila: int
    datos: dict[str, str]


class ImportPreviewResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    preview_token: str
    materia_id: UUID
    cohorte_id: UUID
    actividades: list[DetectedActivity]
    total_rows: int
    alumnos_match: list[StudentMatch]
    alumnos_no_match: list[UnmatchedRow]


class ImportConfirmRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    preview_token: str
    actividad_ids: list[str]


class ImportConfirmResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    materia_id: UUID
    cohorte_id: UUID
    registros_creados: int
    actividades_importadas: list[str]


# ── Completion Report ───────────────────────────────────────────


class PosibleEntregaSinCorregir(BaseModel):
    model_config = ConfigDict(extra="forbid")

    alumno_nombre: str
    alumno_apellidos: str
    actividad: str


class CompletionReportResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    materia_id: UUID
    cohorte_id: UUID
    posibles_entregas_sin_corregir: list[PosibleEntregaSinCorregir]

File: frontend/src/features/comisiones/types/calificaciones.ts
export interface Calificacion {
  id: string
  tenant_id: string
  entrada_padron_id: string
  materia_id: string
  actividad: string
  nota_numerica: number | null
  nota_textual: string | null
  aprobado: boolean
  origen: string
  importado_at: string | null
  created_at: string
  updated_at: string
  deleted_at: string | null
}

export interface CalificacionesResponse {
  items: Calificacion[]
  total: number
}

export interface ActividadDetectada {
  nombre: string
  tipo: string
}

export interface ImportPreviewResponse {
  preview_token: string
  materia_id: string
  cohorte_id: string
  actividades: ActividadDetectada[]
  total_rows: number
  alumnos_match: Array<{
    entrada_padron_id: string
    nombre: string
    apellidos: string
    email: string
    datos: Record<string, string | number | null>
  }>
  alumnos_no_match: Array<{
    fila: number
    datos: Record<string, string>
  }>
}

export interface ImportConfirmResponse {
  materia_id: string
  cohorte_id: string
  registros_creados: number
  actividades_importadas: string[]
}

export interface UmbralConfig {
  id: string
  tenant_id: string
  asignacion_id: string
  materia_id: string
  umbral_pct: number
  valores_aprobatorios: string[] | null
  created_at: string
  updated_at: string
  deleted_at: string | null
}

export interface AtrasadoItem {
  entrada_padron_id: string
  alumno_nombre: string
  actividades_atrasadas: Array<{
    actividad: string
    motivo: string
  }>
}

export interface RankingItem {
  ranking: number
  entrada_padron_id: string
  alumno_nombre: string
  actividades_aprobadas: number
}

export interface RankingResponse {
  items: RankingItem[]
  total: number
}

export interface NotaFinalItem {
  entrada_padron_id: string
  alumno_nombre: string
  promedio: number
  aprobado: boolean
}

export interface NotasFinalesResponse {
  items: NotaFinalItem[]
  total: number
}

export interface ReportesRapidosResponse {
  total_alumnos: number
  total_calificaciones: number
  promedio_general: number
  total_aprobados: number
  total_no_aprobados: number
  desglose_por_actividad: Array<{
    actividad: string
    presentado: number
    promedio: number | null
    min: number | null
    max: number | null
  }>
}
